include/cxxeffect/task.hpp

```diff
@@ -0,0 +1,53 @@
+#pragma once
+#include <functional>
+#include <future>
+
+namespace eff {
+    template <class F, class T>
+    using flatmap_t = decltype(std::declval<F>()(std::declval<T>()).unsafeRunSync());
```

Comment:
Seems wrong to use `unsafeRunSync()` to get the type out of the type constructor. Maybe we could add a type-member to `task`? Would that work?

Reply:
I don't think this is necessary at all, its just ```T```, right?

Reply:
`F` is _like_ a `T => task<U>` where `U` needs to be derived from `F`

Reply:
We could add a type member to task! Then it would look like:
```cpp
template <class F, class T>
using flatmap_t = typename decltype(std::declval<F>()(std::declval<T>()))::type;
```

Reply:
> I don't think this is necessary at all, its just `T`, right?

It's not `T`. `flatmap_t` is the type returned by the task generated by invoking a function of type `F` on an input of type `T`. So if `F` is `int -> task<std::string>`, then `flatmap_t<F, int>` is `std::string`

Reply:
Let's add a type member then. What would std call it? `result_t`?

Reply:
Yeah I see now, it is necessary. Because flatmap takes ```class F``` which is not necessarily an ```std::function```, so you need to determine what type it produces. And it can't be a member of the class, because it changes based on ```class F```. Very nice!